### PR TITLE
Chapter 12: Reset password

### PR DESCRIPTION
**Pull request by stephanie1997-kek**

### CODE REVIEW

**Comment by stephanie1997-kek** [comment]
ready

File: app/assets/stylesheets/password_resets.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/password_resets.scss]: remove file that never used

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  	if @user
+  	  @user.create_reset_digest
+  	  @user.send_password_reset_email
+  	  flash[:info] = t ".success_acct_sent"
+  	  redirect_to root_path
+  	else
+  	  flash.now[:danger] = t ".fail_acct_sent"
+  	  render :new
+  	end
+  end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: check indenting, please

File: app/helpers/password_resets_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/password_resets_helper.rb]: remove file that never used

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
     @user = user
     mail to: user.email, subject: t(".acct_active")
   end
 
-  def password_reset
-    @greeting = t ".greeting"
-    mail to: "to@example.org"
+  def password_reset user
+  	@user = user
+  	mail to: user.email, subject: t(".ps_reset_mess")
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: indent

File: app/views/password_resets/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  	<%= form_for :password_reset, url: password_resets_new_path do |f| %>
+  	  <%= f.label :email %>
+  	  <%= f.email_field :email, class: "form-control" %>
+  	  <%= f.submit t(".forgot_submit"), class: "btn btn-primary" %>
+  	<% end %>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb]: add endline, please

File: config/routes.rb [hunk 6]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
 
+  get 'password_resets/new'
+  get 'password_resets/edit'
```
Comment (saobv-2223) [inline config/routes.rb]: using " and move them into scope

File: config/settings.yml [hunk 7]
```diff
@@ -27,6 +27,7 @@ collections:
     size:
       default: 80
       list: 50
+  psword_reset_expired_time_in_hours: 2
 gravartar:
   size_default: 80
```
Comment (saobv-2223) [inline config/settings.yml]: endline

File: app/assets/stylesheets/password_resets.scss (hunk 1 above)
Comment (stephanie1997-kek) [reply app/assets/stylesheets/password_resets.scss]: done

File: app/controllers/password_resets_controller.rb (hunk 2 above)
Comment (stephanie1997-kek) [reply app/controllers/password_resets_controller.rb]: done

File: app/helpers/password_resets_helper.rb (hunk 3 above)
Comment (stephanie1997-kek) [reply app/helpers/password_resets_helper.rb]: done

File: app/mailers/user_mailer.rb (hunk 4 above)
Comment (stephanie1997-kek) [reply app/mailers/user_mailer.rb]: done

File: app/views/password_resets/new.html.erb (hunk 5 above)
Comment (stephanie1997-kek) [reply app/views/password_resets/new.html.erb]: done

**Comment by stephanie1997-kek** [comment]
ready.

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    if @user
+  	  @user.create_reset_digest
+  	  @user.send_password_reset_email
+  	  flash[:info] = t ".success_acct_sent"
+  	  redirect_to root_path
+    else
+  	  flash.now[:danger] = t ".fail_acct_sent"
+  	  render :new
+    end
+  end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: indenting only 2 spaces, please

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def load_user
+    @user = User.find_by email: params[:email]
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:44]:
```suggestion
    @user = User.find_by email: params[:email]
    return if @user
    
    flash[:danger] = "Some messages"
    redirect_to some_path
```

File: app/mailers/user_mailer.rb [hunk 10]
```diff
@@ -1,11 +1,12 @@
...
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: indenting only 2 spaces, please

File: app/views/password_resets/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+<h1><%= t ".heading" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  	<%= form_for :password_reset, url: password_resets_new_path do |f| %>
+  	  <%= f.label :email %>
+  	  <%= f.email_field :email, class: "form-control" %>
+  	  <%= f.submit t(".forgot_submit"), class: "btn btn-primary" %>
+  	<% end %>
+  </div>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb]: indenting

File: config/settings.yml [hunk 12]
```diff
@@ -23,10 +23,12 @@ validations:
...
 collections:
   session_controller_create: 1
+  psword_reset_expired_time_in_hours: 2
   gravatar:  
     size:
       default: 80
       list: 50
+
 gravartar:
   size_default: 80
```
Comment (saobv-2223) [inline config/settings.yml]:
```suggestion
  
  
```

File: app/controllers/password_resets_controller.rb (hunk 9 above)
Comment (stephanie1997-kek) [reply app/controllers/password_resets_controller.rb:44]: done

File: app/mailers/user_mailer.rb (hunk 10 above)
Comment (stephanie1997-kek) [reply app/mailers/user_mailer.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+  def new; end
+
+  def create
+    @user = User.find_by email: params[:password_reset][:email].downcase
+    if @user
+  	 @user.create_reset_digest
+     @user.send_password_reset_email
+     flash[:info] = t ".success_acct_sent"
+     redirect_to root_path
```
Comment (manhnt-0960) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      @user.create_reset_digest
      @user.send_password_reset_email
      flash[:info] = t ".success_acct_sent"
      redirect_to root_path
```

File: app/controllers/password_resets_controller.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  	 flash.now[:danger] = t ".fail_acct_sent"
+  	 render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
+      @user.errors.add :password, "can't be empty"
```
Comment (manhnt-0960) [inline app/controllers/password_resets_controller.rb]: please add i18n

File: app/models/user.rb [hunk 15]
```diff
@@ -57,6 +57,15 @@ def send_activation_email
     UserMailer.account_activation(self).deliver_now
   end
 
+  def create_reset_digest
```
Comment (manhnt-0960) [inline app/models/user.rb]: let's move this method into private

File: config/routes.rb [hunk 16]
```diff
@@ -11,7 +11,10 @@
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     delete "/logout", to: "sessions#destroy"
+    get "password_resets/new"
+    get "password_resets/edit"
```
Comment (manhnt-0960) [inline config/routes.rb]: let's remove them, because u use `resources :password_resets`  below

File: config/settings.yml [hunk 17]
```diff
@@ -23,10 +23,13 @@ validations:
...
-  
\ No newline at end of file
+
+
```
Comment (manhnt-0960) [inline config/settings.yml:34]: remove empty line

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    else
+  	 flash.now[:danger] = t ".fail_acct_sent"
+  	 render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    if params[:user][:password].blank?
```

File: config/settings.yml [hunk 19]
```diff
@@ -23,10 +23,12 @@ validations:
...
   gravatar:  
     size:
       default: 80
       list: 50
+
 gravartar:
   size_default: 80
-  
\ No newline at end of file
+
```
Comment (saobv-2223) [inline config/settings.yml:34]:
```suggestion
```

Comment (stephanie1997-kek) [reply config/settings.yml:34]: but this is the end line for this file. Do I have to delete it too?

**Comment by stephanie1997-kek** [comment]
ready

File: app/views/password_resets/new.html.erb [hunk 20]
```diff
@@ -0,0 +1,12 @@
+<% provide :title, t(".forgot_title")  %>
+<h1><%= t ".heading" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+   <%= form_for :password_reset, url: password_resets_new_path do |f| %>
+    <%= f.label :email %>
+  	<%= f.email_field :email, class: "form-control" %>
+  	<%= f.submit t(".forgot_submit"), class: "btn btn-primary" %>
+   <% end %>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb]: check indenting

File: config/settings.yml (hunk 19 above)
Comment (saobv-2223) [inline config/settings.yml:34]: image

**Comment by stephanie1997-kek** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**